File: WG.m
function [Res_1,Res_0] = WG(log,wgxaxis,wgyaxis,currentWG0,currentWG1,logvars,plot,WGlogic)

%% Determine SWG/FF

% WGlogic = questdlg('Are you using FeedForward or SWG?','WG Logic','FF','SWG','FF');

if WGlogic==1
    log.EFF=log.RPM
    log.IFF=log.PUTSP*10
%     wgyaxis=wgyaxis/10
end

    oldwgyaxis = wgyaxis
    oldwgxaxis = wgxaxis   

%% Get other inputs

prompt = {'PUT fudge factor:','Minimum pedal (if no PUT_I_INHIBIT):','Maximum PUT delta:','Minimum Boost:'};
dlgtitle = 'WG Inputs';
dims = [1 50];
definput = {'0.71','50','10','0'};
answer = inputdlg(prompt,dlgtitle,dims,definput)
fudge=str2num(answer{1});
minpedal=str2num(answer{2});
maxdelta=str2num(answer{3});
minboost=str2num(answer{4});

%% Create Derived Values
log.deltaPUT=log.PUT-log.PUTSP;
log.WGNEED=log.WG_Final-log.deltaPUT.*fudge;
log.WGCL=log.WG_Final-log.WG_Base;

%% Create Trimmed datasets

log=log;

if any(contains(logvars,'I_INH'))
    log(log.I_INH>0,:) = [];
else
    log(log.Pedal<minpedal,:) = [];
end

if any(contains(logvars,'DV'))
    log(log.DV>50,:) = [];
end

if any(contains(logvars,'BOOST'))
    log(log.BOOST<minboost,:) = [];
end

log(abs(log.deltaPUT)>maxdelta,:) = [];
log(log.WG_Final>98,:) = [];
log_WGopen=log;


%% Plot
if plot==true
    wait=waitbar(0,"Plotting")
    for i=1:height(log)
        pts(i,:)=[log.EFF(i),log.IFF(i),100,log.deltaPUT(i)];
        if log.VVL(i)==1
            syms(i)="^";
        else
            syms(i)="o";
        end
        waitbar(i/height(log),wait,"Choosing Points");
    end    
    hold on

    mycolormap = customcolormap([0 .25 .5 .75 1], {'#9d0142','#f66e45','#ffffff','#65c0ae','#5e4f9f'});
    c = colorbar;
    c.Label.String = 'PUT - PUT SP';
    colormap(mycolormap);
    clim([-15 15]);
    set(gca, 'Ydir', 'reverse');
    xlabel('Exh Flow Factor') ;
    ylabel('Int Flow Factor') ;
    grid on;
    set(gca,"XAxisLocation","top");
    xticks(wgxaxis);
    yticks(wgyaxis);

    for i=1:height(log)
        if mod(i,ceil(height(log)/1000))==0
            scatr=scatter(pts(i,1),pts(i,2),pts(i,3),pts(i,4),syms(i));
        end
        if mod(i,ceil(height(log)/100))==0
            waitbar(i/height(log),wait,"Plotting Points"); 
        end
    end
    close(wait)
    
    figedit = uifigure('Name','DO NOT CLOSE THIS TABLE USING THE X, USE CONTINUE BUTTON',"Position",[500 500 760 320]);
    lbl2 = uilabel(figedit,"Text",'WG X Axis','Position',[20 290 720 30]);
    uit2 = uitable(figedit, "Data",wgxaxis, "Position",[20 210 720 80], 'ColumnEditable',true);
    lbl3 = uilabel(figedit,"Text",'WG Y Axis','Position',[20 150 720 30]);
    uit3 = uitable(figedit, "Data",wgyaxis, "Position",[20 70 720 80], 'ColumnEditable',true);
    btn1 = uicontrol(figedit,'String','Replot','Position',[20 30 150 30])
    btn1.Callback={@plotButtonPushed,uit2,uit3}
    c = uicontrol(figedit,'String','Continue','Position',[220 30 150 30])
    c.Callback={@cb_continue,figedit}

    uiwait(figedit)
%     hold off
    oldwgyaxis = wgyaxis
    oldwgxaxis = wgxaxis   
    wgxaxis = uit2.Data
    wgyaxis = uit3.Data
    close(figedit)
end

%% Create Bins

wgxedges(1)=wgxaxis(1);
wgxedges(length(wgxaxis)+1)=wgxaxis(length(wgxaxis))+2;
for i=1:length(wgxaxis)-1;
    wgxedges(i+1)=(wgxaxis(i)+wgxaxis(i+1))/2;
end

wgyedges(1)=wgyaxis(1);
wgyedges(length(wgyaxis)+1)=wgyaxis(length(wgyaxis))+2;
for i=1:length(wgyaxis)-1;
    wgyedges(i+1)=(wgyaxis(i)+wgyaxis(i+1))/2;
end


exhlabels=string(wgxaxis);
intlabels=string(wgyaxis);

log_WGopen.X=discretize(log_WGopen.EFF,wgxedges);
log_WGopen.Y=discretize(log_WGopen.IFF,wgyedges);


log_VVL1=log_WGopen;
log_VVL1(log_VVL1.VVL~=1,:) = [];
log_VVL0=log_WGopen;
log_VVL0(log_VVL0.VVL~=0,:) = [];

%% Initialize matrixes

SUM1=zeros(length(wgyaxis),length(wgxaxis));
COUNT1=SUM1;
SUM0=SUM1;
COUNT0=SUM1;
COUNTFAC=zeros(4,4);
SUMFAC=zeros(4,4);
COUNTOFS=zeros(4,4);
SUMOFS=zeros(4,4);
columns1= zeros(10,16)
rows1=columns1
rows0=rows1
columns0=columns1

%% Discretize VVL1

for i=1:length(wgxaxis)
        temp=log_VVL1;
        temp(temp.X~=i,:)=[];
        if height(temp)>1
            columns1(:,i) = lsq_lut_piecewise( temp.IFF, temp.WGNEED, wgyaxis )
        end
end

for j=1:length(wgyaxis)
        temp=log_VVL1;
        temp(temp.Y~=j,:)=[];
        if height(temp)>1
            rows1(j,:) = lsq_lut_piecewise( temp.EFF, temp.WGNEED, wgxaxis )
        end
end

blend1=(columns1+rows1)/200

for i=1:length(wgxaxis)
    for j=1:length(wgyaxis)
        temp=log_VVL1;
        temp(temp.X~=i,:)=[];
        temp(temp.Y~=j,:)=[];
        AVGtemp=mean(temp.WGNEED)
        COUNT(j,i)=height(temp)
        current(j,i)=interp2(oldwgxaxis,oldwgyaxis,currentWG1,wgxaxis(i),wgyaxis(j),"linear",0)
            if COUNT(j,i)>3
            ci=paramci(fitdist(temp.WGNEED,'Normal'),'Alpha',0.5)
            sigma(j,i)=ci(2,2)
            low(j,i)=ci(1,1)
            high(j,i)=ci(2,1)
%             AVGtemp=0
            if isnan(current(j,i))
                AVG1(j,i)=(blend1(j,i)+AVGtemp/100)/2
            elseif low(j,i)>current(j,i)
                AVG1(j,i)=(blend1(j,i)+AVGtemp/100)/2
            elseif high(j,i)<current(j,i)
                AVG1(j,i)=(blend1(j,i)+AVGtemp/100)/2
            else
                AVG1(j,i)=current(j,i)
            end
        else
            AVG1(j,i)=current(j,i)
        end
    end
end

AVG1=round(AVG1*16384)/16384

Data1=current-AVG1
Data1=Data1./Data1

Res_1=array2table(AVG1,'VariableNames',exhlabels,'RowNames',intlabels);

%% Discretize VVL0

for i=1:length(wgxaxis)
        temp=log_VVL0;
        temp(temp.X~=i,:)=[];
        if height(temp)>1
            columns0(:,i) = lsq_lut_piecewise( temp.IFF, temp.WGNEED, wgyaxis)
        end
end

for j=1:length(wgyaxis)
        temp=log_VVL0;
        temp(temp.Y~=j,:)=[];
        if height(temp)>1   
            rows0(j,:) = lsq_lut_piecewise( temp.EFF, temp.WGNEED, wgxaxis )  
        end
end

blend0=(columns0+rows0)/200

for i=1:length(wgxaxis)
    for j=1:length(wgyaxis)
        temp=log_VVL0;
        temp(temp.X~=i,:)=[];
        temp(temp.Y~=j,:)=[];
        AVGtemp=mean(temp.WGNEED)
        COUNT(j,i)=height(temp)
        current(j,i)=interp2(oldwgxaxis,oldwgyaxis,currentWG0,wgxaxis(i),wgyaxis(j))
            if COUNT(j,i)>3
            ci=paramci(fitdist(temp.WGNEED,'Normal'),'Alpha',0.5)
            sigma(j,i)=ci(2,2)
            low(j,i)=ci(1,1)
            high(j,i)=ci(2,1)
%             AVGtemp=0
            if isnan(current(j,i))
                AVG0(j,i)=(blend0(j,i)+AVGtemp/100)/2                
            elseif low(j,i)>current(j,i)
                AVG0(j,i)=(blend0(j,i)+AVGtemp/100)/2
            elseif high(j,i)<current(j,i)
                AVG0(j,i)=(blend0(j,i)+AVGtemp/100)/2
            else
                AVG0(j,i)=current(j,i)
            end
        else
            AVG0(j,i)=current(j,i)
        end
    end
end

AVG0=round(AVG0*16384)/16384

Data0=current-AVG0
Data0=Data0./Data0

Res_0=array2table(AVG0,'VariableNames',exhlabels,'RowNames',intlabels);


% %% Plot Scatters
% if plot == 1
%     f1=tiledlayout(2,1)
%     nexttile
%     hold on
%     for i=1:height(log)
%         if log.VVL(i)==1
%             if log.WG_Final(i)>98
%                 c1=scatter(log.EFF(i),log.IFF(i),100,log.deltaPUT(i),"^","filled");
%             else
%                 o1=scatter(log.EFF(i),log.IFF(i),100,log.deltaPUT(i),"^");
%             end
%         else
%             if log.WG_Final(i)>98
%                 c0=scatter(log.EFF(i),log.IFF(i),100,log.deltaPUT(i),"o","filled");
%             else
%                 o0=scatter(log.EFF(i),log.IFF(i),100,log.deltaPUT(i),"o");
%             end
%         end
%     end
%     
%     mycolormap = customcolormap([0 .25 .5 .75 1], {'#9d0142','#f66e45','#ffffff','#65c0ae','#5e4f9f'});
%     c = colorbar;
%     c.Label.String = 'PUT - PUT SP';
%     colormap(mycolormap);
%     clim([-15 15]);
%     set(gca, 'Ydir', 'reverse');
%     xlabel('Exh Flow Factor') ;
%     ylabel('Int Flow Factor') ;
%     % set(gca,'TickLength',[1 1])
%     grid on;
%     set(gca,"XAxisLocation","top");
%     xticks(wgxaxis);
%     yticks(wgyaxis);
%     rows=[1,1,1,1]
%     rowsfound=[find(log.RPM>2000,1) find(log.RPM>3000,1) find(log.RPM>4000,1) find(log.RPM>5000,1)];
%     rows(1,1:numel(rowsfound)) = rowsfound;
%     lscatter(log.EFF(rows),log.IFF(rows),[2000 3000 4000 5000]);
%     hold off
%     
%     nexttile
%     hold on
%     for i=1:height(log)
%         if log.VVL(i)==1
%             if log.WG_Final(i)>98
%                 c1p=scatter(log.EFF(i),log.IFF(i),100,log.WGCL(i),"^","filled");
%             else
%                 o1p=scatter(log.EFF(i),log.IFF(i),100,log.WGCL(i),"^");
%             end
%         else
%             if log.WG_Final(i)>98
%                 c0p=scatter(log.EFF(i),log.IFF(i),100,log.WGCL(i),"o","filled");
%             else
%                 o0p=scatter(log.EFF(i),log.IFF(i),100,log.WGCL(i),"o");
%             end
%         end
%     end
%     
%     mycolormap = customcolormap([0 .25 .5 .75 1], {'#9d0142','#f66e45','#ffffff','#65c0ae','#5e4f9f'});
%     c = colorbar;
%     c.Label.String = 'WG CL';
%     colormap(mycolormap);
%     clim([-10 10]);
%     set(gca, 'Ydir', 'reverse');
%     xlabel('Exh Flow Factor') ;
%     ylabel('Int Flow Factor') ;
%     % set(gca,'TickLength',[1 1])
%     grid on;
%     set(gca,"XAxisLocation","top");
%     xticks(wgxaxis);
%     yticks(wgyaxis);
%     rows=[1,1,1,1]
%     rowsfound=[find(log.RPM>2000,1) find(log.RPM>3000,1) find(log.RPM>4000,1) find(log.RPM>5000,1)];
%     rows(1,1:numel(rowsfound)) = rowsfound;
%     lscatter(log.EFF(rows),log.IFF(rows),[2000 3000 4000 5000]);
%     hold off
% end
% 

%% Display Tables

fig0 = uifigure('Name','VVL0 WG Results',"Position",[500 500 760 360]);
uit = uitable(fig0, "Data",Res_0, "Position",[20 20 720 320]);

styleIndices = ~ismissing(Data0);
[row,col] = find(styleIndices);
s = uistyle("BackgroundColor",'gr');
addStyle(uit,s,"cell",[row,col]);

fig1 = uifigure('Name','VVL1 WG Results',"Position",[500 500 760 360]);
uit = uitable(fig1, "Data",Res_1, "Position",[20 20 720 320]);

styleIndices = ~ismissing(Data1);
[row,col] = find(styleIndices);
s = uistyle("BackgroundColor",'gr');
addStyle(uit,s,"cell",[row,col]);
end

function plotButtonPushed(src,event,uit2,uit3)
    xticks(uit2.Data)
    yticks(uit3.Data)
end

function cb_continue(~,~,figedit)
    uiresume(figedit);
end

File: WG.py
# C:/Users/Sam/PycharmProjects/YAKtunerCONVERTED/WG.py

"""
Wastegate (WG) Tuning Module for YAKtuner

This module contains pure, non-UI functions to analyze engine log data and
recommend adjustments to wastegate base tables.
"""

import numpy as np
import pandas as pd
from scipy import stats, interpolate
import matplotlib.pyplot as plt

# --- Core Calculation and Filtering Functions ---

def _process_and_filter_log_data(log_df, params, logvars, WGlogic, tempcomp, tempcompaxis, min_pedal=50.0):
    """
    A pure function to prepare and filter log data without UI interactions.
    """
    warnings = []
    processed_log = log_df.copy()

    # Determine SWG/FF logic and apply temperature correction
    if WGlogic:
        processed_log['EFF'] = processed_log['RPM']
        processed_log['IFF'] = processed_log['PUTSP'] * 10
        interp_func = interpolate.interp1d(tempcompaxis, tempcomp, kind='linear', fill_value='extrapolate')
        tempcorr = interp_func(processed_log['AMBTEMP'])
    else:
        tempcorr = 0

    # Create derived values for analysis
    processed_log['deltaPUT'] = processed_log['PUT'] - processed_log['PUTSP']
    processed_log['WGNEED_uncorrected'] = processed_log['WG_Final'] - processed_log['deltaPUT'] * params['fudge']
    processed_log['WGNEED'] = processed_log['WGNEED_uncorrected'] - tempcorr

    # Filter log data to valid conditions
    if 'I_INH' in logvars:
        processed_log = processed_log[processed_log['I_INH'] <= 0]
    else:
        warnings.append("Recommend logging 'PUT I Inhibit'. Using pedal position as a fallback.")
        processed_log = processed_log[processed_log['Pedal'] >= min_pedal]

    if 'DV' in logvars:
        processed_log = processed_log[processed_log['DV'] <= 50]
    else:
        warnings.append("Recommend logging 'DV position'. Otherwise, DV may impact accuracy.")

    if 'BOOST' in logvars:
        processed_log = processed_log[processed_log['BOOST'] >= params['minboost']]
    else:
        warnings.append("Recommend logging 'boost'. Otherwise, logs are not trimmed for min boost.")

    # Filtering Logic for PUT Delta stability
    processed_log['deltaPUT_CHANGE'] = processed_log['deltaPUT'].diff().abs()
    is_small_delta = processed_log['deltaPUT'].abs() < 10
    is_steady_delta = processed_log['deltaPUT_CHANGE'] < 1.0
    final_mask = is_small_delta | is_steady_delta
    processed_log = processed_log[final_mask.fillna(False)]

    if processed_log.empty:
        warnings.append("No data points met the criteria (small or steady PUT delta).")

    # Final filter for WG duty cycle range
    processed_log = processed_log[processed_log['WG_Final'] <= 98]

    return processed_log, warnings

def _create_bins_and_labels(log_df, wgxaxis, wgyaxis):
    """Creates bin edges from axes and assigns each log entry to a grid cell (X, Y)."""
    wgxedges = np.zeros(len(wgxaxis) + 1)
    wgxedges[0] = wgxaxis[0]
    wgxedges[-1] = wgxaxis[-1] + 2
    for i in range(len(wgxaxis) - 1):
        wgxedges[i + 1] = (wgxaxis[i] + wgxaxis[i + 1]) / 2

    wgyedges = np.zeros(len(wgyaxis) + 1)
    wgyedges[0] = wgyaxis[0]
    wgyedges[-1] = wgyaxis[-1] + 2
    for i in range(len(wgyaxis) - 1):
        wgyedges[i + 1] = (wgyaxis[i] + wgyaxis[i + 1]) / 2

    log_df['X'] = pd.cut(log_df['EFF'], wgxedges, labels=False)
    log_df['Y'] = pd.cut(log_df['IFF'], wgyedges, labels=False)
    return log_df

def create_wg_scatter_plot(log_VVL0, log_VVL1, wgxaxis, wgyaxis, WGlogic):
    """
    Creates a Matplotlib scatter plot figure of the filtered log data.
    """
    fig, ax = plt.subplots(figsize=(12, 8))
    scatter1 = ax.scatter(
        log_VVL1['EFF'], log_VVL1['IFF'], s=abs(log_VVL1['WGNEED']),
        c=log_VVL1['deltaPUT'], marker='x', cmap='RdBu', label='VVL1'
    )
    ax.scatter(
        log_VVL0['EFF'], log_VVL0['IFF'], s=abs(log_VVL0['WGNEED']),
        c=log_VVL0['deltaPUT'], marker='o', cmap='RdBu', label='VVL0', alpha=0.7
    )
    cbar = fig.colorbar(scatter1, ax=ax)
    cbar.set_label('PUT - PUT SP (kPa)')
    ax.invert_yaxis()
    if WGlogic:
        ax.set_xlabel('RPM')
        ax.set_ylabel('PUT SP')
    else:
        ax.set_xlabel('Engine Efficiency (EFF)')
        ax.set_ylabel('Intake Flow Factor (IFF)')
    ax.set_title('Wastegate Duty Cycle Need vs. Operating Point')
    ax.grid(True)
    ax.set_xticks(wgxaxis)
    ax.set_xticklabels(labels=wgxaxis, rotation=45)
    ax.set_yticks(wgyaxis)
    ax.legend()
    fig.tight_layout()
    return fig

def _fit_surface(log_data, wgxaxis, wgyaxis):
    """
    Fits a 3D surface to the provided log data using scipy.interpolate.griddata.
    """
    if log_data.empty or len(log_data) < 3:
        return np.zeros((len(wgyaxis), len(wgxaxis)))
    points = log_data[['EFF', 'IFF']].values
    values = log_data['WGNEED'].values
    grid_x, grid_y = np.meshgrid(wgxaxis, wgyaxis)
    fitted_surface = interpolate.griddata(points, values, (grid_x, grid_y), method='linear')
    nan_mask = np.isnan(fitted_surface)
    if np.any(nan_mask):
        nearest_fill = interpolate.griddata(points, values, (grid_x[nan_mask], grid_y[nan_mask]), method='nearest')
        fitted_surface[nan_mask] = nearest_fill
    if np.all(np.isnan(fitted_surface)):
        return np.zeros((len(wgyaxis), len(wgxaxis)))
    return fitted_surface / 100.0

def _calculate_final_recommendations(log_data, blend, old_table, wgxaxis, wgyaxis, calculate_temp_coef=False):
    """
    Calculates the final recommended WG table by comparing the new fit with the old
    table and applying confidence intervals to each cell.
    """
    final_table = old_table.copy()
    total_coef, count_coef = 0, 0
    for i in range(len(wgxaxis)):
        for j in range(len(wgyaxis)):
            cell_data = log_data[(log_data['X'] == i) & (log_data['Y'] == j)]
            if len(cell_data) > 3:
                mean, std_dev = stats.norm.fit(cell_data['WGNEED'])
                low_ci, high_ci = stats.norm.interval(0.7, loc=mean, scale=std_dev)
                current_val_scaled = old_table[j, i] * 100.0
                if np.isnan(current_val_scaled) or not (low_ci <= current_val_scaled <= high_ci):
                    new_val = (blend[j, i] + (mean / 100.0)) / 2
                    final_table[j, i] = new_val
            if calculate_temp_coef:
                temp_log = cell_data[cell_data['BOOST'] >= 8]
                if len(temp_log) > 2 and (temp_log['AMBTEMP'].max() > temp_log['AMBTEMP'].min() + 15):
                    coef = np.polyfit(temp_log['AMBTEMP'], temp_log['WGNEED_uncorrected'] / 100, 1)[0]
                    total_coef += coef * len(temp_log)
                    count_coef += len(temp_log)
    final_table = np.round(final_table * 16384) / 16384
    if calculate_temp_coef:
        avg_coef = total_coef / count_coef if count_coef > 0 else None
        return final_table, avg_coef
    else:
        return final_table

# --- Main Orchestrator Function ---

def run_wg_analysis(log_df, wgxaxis, wgyaxis, oldWG0, oldWG1, logvars, WGlogic, tempcomp, tempcompaxis, show_scatter_plot=True):
    """
    Main orchestrator for the WG tuning process. A pure computational function.

    Args:
        log_df (pd.DataFrame): The mapped log data.
        wgxaxis, wgyaxis (np.ndarray): The axes for the WG tables.
        oldWG0, oldWG1 (np.ndarray): The original WG tables from the tune.
        logvars (list): A list of available variable names in the log.
        WGlogic (bool): Flag for SWG/FF logic.
        tempcomp, tempcompaxis (np.ndarray): Temperature compensation data.
        show_scatter_plot (bool): Flag to control generation of the scatter plot.

    Returns:
        dict: A dictionary containing all results:
              - 'status' (str): 'Success' or 'Failure'.
              - 'warnings' (list): A list of warning messages.
              - 'scatter_plot_fig' (matplotlib.figure.Figure or None): The generated plot.
              - 'results_vvl0' (pd.DataFrame or None): The recommended VVL0 table.
              - 'results_vvl1' (pd.DataFrame or None): The recommended VVL1 table.
              - 'temp_comp_results' (pd.DataFrame or None): Temp comp recommendations.
    """
    print(" -> Initializing WG analysis...")
    # Parameters are now hardcoded here, but could be passed in from the UI.
    params = {'fudge': 0.71, 'minboost': 0}

    print(" -> Preparing and filtering log data...")
    processed_log, warnings = _process_and_filter_log_data(
        log_df=log_df, params=params, logvars=logvars, WGlogic=WGlogic,
        tempcomp=tempcomp, tempcompaxis=tempcompaxis
    )

    if processed_log.empty:
        return {'status': 'Failure', 'warnings': warnings, 'scatter_plot_fig': None,
                'results_vvl0': None, 'results_vvl1': None, 'temp_comp_results': None}

    print(" -> Creating data bins from WG axes...")
    log = _create_bins_and_labels(processed_log, wgxaxis, wgyaxis)

    print(" -> Separating data for VVL0 and VVL1...")
    log_VVL1 = log[log['VVL'] == 1].copy()
    log_VVL0 = log[log['VVL'] == 0].copy()

    scatter_fig = None
    if show_scatter_plot:
        print(" -> Generating raw WG data plot...")
        scatter_fig = create_wg_scatter_plot(log_VVL0, log_VVL1, wgxaxis, wgyaxis, WGlogic)

    print(" -> Fitting 3D surface for VVL1...")
    blend1 = _fit_surface(log_VVL1, wgxaxis, wgyaxis)

    print(" -> Fitting 3D surface for VVL0...")
    blend0 = _fit_surface(log_VVL0, wgxaxis, wgyaxis)

    print(" -> Calculating final recommendations for VVL1...")
    final_table_1 = _calculate_final_recommendations(log_VVL1, blend1, oldWG1, wgxaxis, wgyaxis)

    print(" -> Calculating final recommendations for VVL0 and Temp Comp...")
    final_table_0, avg_coef = _calculate_final_recommendations(
        log_VVL0, blend0, oldWG0, wgxaxis, wgyaxis, calculate_temp_coef=True
    )

    # NOTE: 3D plotting logic is removed from here. It should be handled in the
    # Streamlit script if desired, using a refactored `plot_3d_surface` function.

    print(" -> Preparing temperature compensation results...")
    temp_comp_results_df = None
    if avg_coef is not None:
        _slope, original_intercept = np.polyfit(tempcompaxis, tempcomp, 1)
        new_tempcomp = (avg_coef * tempcompaxis) + original_intercept
        temp_df = pd.DataFrame({
            'Temperature': tempcompaxis,
            'Original Comp': tempcomp,
            'Recommended Comp': new_tempcomp
        })
        temp_comp_results_df = temp_df.set_index('Temperature').T.round(4)

    print(" -> Preparing final results as DataFrames...")
    exhlabels = [str(x) for x in wgxaxis]
    intlabels = [str(x) for x in wgyaxis]
    Res_1 = pd.DataFrame(final_table_1, columns=exhlabels, index=intlabels)
    Res_0 = pd.DataFrame(final_table_0, columns=exhlabels, index=intlabels)

    print(" -> WG analysis complete.")
    return {
        'status': 'Success',
        'warnings': warnings,
        'scatter_plot_fig': scatter_fig,
        'results_vvl0': Res_0,
        'results_vvl1': Res_1,
        'temp_comp_results': temp_comp_results_df
    }